keyvi/include/keyvi/index/internal/merge_job.h

```diff
@@ -0,0 +1,148 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+#ifndef KEYVI_INDEX_INTERNAL_MERGE_JOB_H_
+#define KEYVI_INDEX_INTERNAL_MERGE_JOB_H_
+
+#include <atomic>
+#include <chrono>  //NOLINT
+#include <functional>
+#include <memory>
+#include <string>
+#include <thread>  // NOLINT
+#include <vector>
+
+#include "process.hpp"
+
+#include "dictionary/dictionary_merger.h"
+#include "dictionary/dictionary_types.h"
+#include "dictionary/fsa/internal/json_value_store.h"
+#include "dictionary/fsa/internal/sparse_array_persistence.h"
+#include "index/internal/segment.h"
+
+// #define ENABLE_TRACING
+#include "dictionary/util/trace.h"
+
+namespace keyvi {
+namespace index {
+namespace internal {
+
+class MergeJob final {
+  struct MergeJobPayload {
+    explicit MergeJobPayload(std::vector<segment_t> segments, const boost::filesystem::path& output_filename)
+        : segments_(segments), output_filename_(output_filename), process_finished_(false) {}
+
+    MergeJobPayload(MergeJobPayload&&) = default;
+    MergeJobPayload& operator=(MergeJobPayload&&) = default;
+
+    std::vector<segment_t> segments_;
+    boost::filesystem::path output_filename_;
+    std::chrono::time_point<std::chrono::system_clock> start_time_;
+    std::chrono::time_point<std::chrono::system_clock> end_time_;
+    int exit_code_ = -1;
+    bool merge_done = false;
+    std::atomic_bool process_finished_;
+  };
+
+ public:
+  // todo: add ability to stop merging for shutdown
+  explicit MergeJob(segment_vec_t segments, size_t id, const boost::filesystem::path& output_filename)
+      : payload_(segments, output_filename), id_(id), external_process_() {}
+
+  ~MergeJob() {
+    if (payload_.process_finished_ == false) {
+      EndExternalProcess();
+    }
+  }
+
+  MergeJob(MergeJob&&) = default;
+  MergeJob& operator=(MergeJob&&) = default;
+
+  void Run() { DoExternalProcessMerge(); }
+
+  bool TryFinalize() {
+    // already finished?
+    if (payload_.process_finished_ == true) {
+      return true;
+    }
+    return TryEndExternalProcess();
+  }
+
+  void Finalize() {
+    if (payload_.process_finished_ == false) {
+      EndExternalProcess();
+    }
+  }
+
+  bool Successful() {
+    // todo: handle case when process crashed
+    return payload_.exit_code_ == 0;
+  }
+
+  const std::vector<segment_t>& Segments() const { return payload_.segments_; }
+
+  const segment_t MergedSegment() const {
+    return segment_t(new Segment(payload_.output_filename_, payload_.segments_, false));
+  }
+
+  void SetMerged() { payload_.merge_done = true; }
+
+  const bool Merged() const { return payload_.merge_done; }
+
+  size_t GetId() const { return id_; }
+
+  // todo: ability to kill job/process
+
+ private:
+  MergeJobPayload payload_;
+  size_t id_;
+  std::shared_ptr<TinyProcessLib::Process> external_process_;
+
+  void DoExternalProcessMerge() {
```

Comment:
so FWIW right now only 'external merging' is supported, this is wanted and an important aspect: merging can not crash the indexer, a OOM situation can be handled better (tbd: lower oom weight for merger) as well es priority can be set.

But: if the dictionaries are small, a in-process merge would be better. The idea in the long run is to let the merge policy decide about it.

Reply:
well, hopefully merger is stable enough to be safe from crashes, compiler is also internal, so if there is a bug we have a problem anyhow. If an indexer process crashes it can be restarted and it will continue from where it stopped but we loose all data that were in internal buffers (like the current compiler). This problem is true for any data store. 

At some point in the future I would like to provide a transaction log, that can be plugged. But even then, you need an fsync to be sure that data is backed up.

For this version, data is guaranteed to be backed to disk after calling Flush(), this triggers a compile of the latest data including an implicit fsync at the end.

Reply:
Agree, but in that case if we are "okay" for this kind of bugs to crash it, and then we restart the index. Shouldn't we be okay for just spawning a merger thread and making sure that we are catching all exceptions from that thread ?

--

Yes of course this makes it less safe, but looks like it can be a good tradeoff.

Reply:
the 'internal merger' will spawn a thread and do the merge as part of the thread, it is important that `merge_job` does not block. The interface will be the same so that it is an internal detail whether the merge happens in-process or external.

I haven't started creating safe-guards yet, till now I preferred to let it crash, e.g. the return code isn't checked yet. It's on the list. ;-)